Sources/AWSSDKSwiftCore/Credential/CredentialProvider.swift

```diff
@@ -0,0 +1,61 @@
+//
+//  CredentialProvider.swift
+//  AWSSDKSwift
+//
+//  Created by Adam Fowler on 2019/12/5
+//
+//
+import AWSSigner
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+
+/// Protocol providing future holding a credential
+protocol CredentialProvider {
+    mutating func getCredential() -> EventLoopFuture<Credential>
+}
+
+/// Provides credentials that don't change
+struct StaticCredentialProvider: CredentialProvider {
+    let credential: Credential
+    let eventLoopGroup: EventLoopGroup
+
+    init(credential: Credential,  eventLoopGroup: EventLoopGroup) {
+        self.credential = credential
+        self.eventLoopGroup = eventLoopGroup
+    }
+
+    func getCredential() -> EventLoopFuture<Credential> {
+        return eventLoopGroup.next().makeSucceededFuture(credential)
+    }
+}
+
+/// Provides credentials acquired from the metadata service. These expire and need updated occasionally
+class MetaDataCredentialProvider: CredentialProvider {
+    var credentialFuture: EventLoopFuture<Credential>
+    let httpClient: AWSHTTPClient
+    let lock = Lock()
+
+    init(httpClient: AWSHTTPClient) {
+        let credential = ExpiringCredential(accessKeyId: "", secretAccessKey: "", expiration: Date.init(timeIntervalSince1970: 0))
+        self.credentialFuture = httpClient.eventLoopGroup.next().makeSucceededFuture(credential)
```

Comment:
I'm not a huge fan using the `refreshFuture` all the time. IMHO this future should only be used while refreshing. What is wrong with my proposed solution? It's longer. But in my biased view more readable.

```swift
    func getCredential() -> EventLoopFuture<Credential> {
        self.lock.lock()
        let cred = credential
        self.lock.unlock()
        
        if let cred = cred, cred.nearExpiration() == false {
            // we have credentials and those are still valid
            
            if self.eventLoopGroup is MultiThreadedEventLoopGroup {
              // if we are in a MultiThreadedEventLoopGroup we try to minimize hops.
              return MultiThreadedEventLoopGroup.currentEventLoop!.makeSucceededFuture(cred)
            }
            return self.eventLoopGroup.next().makeSucceededFuture(cred)
        }
        
        // we need to refresh the credentials
        return self.refreshCredentials()
    }
    
    private func refreshCredentials() -> EventLoopFuture<Credential> {
        self.lock.lock()
        defer { self.lock.unlock() }
        
        if let future = credentialFuture {
            // a refresh is already running
            return future
        }
        
        credentialFuture = self.metaDataClient.getMetaData()
            .map { (metadata) -> (Credential) in
                let credential = metadata.credential
                
                self.lock.lock()
                defer { self.lock.unlock() }
                
                self.credentialFuture = nil
                self.credential = credential
                
                return credential
            }

        return credentialFuture!
    }
```

Reply:
- I guess the one thing I didn't like was we were holding two references to the credential, one inside the future and one in the class.
- The other thing was the optional credential, we shouldnt need an optional value there
I will implement the other version and we can compare.